src/mofaflex/_core/datasets/misc.py

```diff
@@ -51,11 +52,39 @@ def __init__(
     ):
         super().__init__()
 
-        self.covariates, self.covariates_names = data.get_covariates(obs_key, obsm_key)
-        self.covariates = {
-            group_name: np.nanmean(np.stack(tuple(group_covars.values()), axis=0), axis=0)
-            for group_name, group_covars in self.covariates.items()
-        }
+        covariates, self.covariates_names = data.get_covariates(obs_key, obsm_key)
+
+        # if data is categorical, get unique categories
+        categories = set()
+        for _, group_covars in covariates.items():
+            for _, view_covars in group_covars.items():
+                if view_covars.dtype == np.object_:
+                    non_nan_values = view_covars[~pd.isnull(view_covars)]
+                    categories.update(set(np.unique(non_nan_values)))
+                    have_nans = non_nan_values.size < view_covars.size
+
+        # map categories to floats
+        categories_mapping = {cat: float(i) if have_nans else i for i, cat in enumerate(sorted(categories))}
+
+        for group_name, group_covars in covariates.items():
+            for view_name, view_covars in group_covars.items():
+                if view_covars.dtype == np.object_:
+                    covariates[group_name][view_name] = np.vectorize(lambda cat: categories_mapping.get(cat, np.nan))(
+                        view_covars
+                    )
+
+        # ensure the a covariate value is consistent across views (nanmean or first)
+        self.covariates = {}
+        for group_name, group_covars in covariates.items():
+            group_covars_stacked = np.stack(tuple(group_covars.values()), axis=0)
+            if np.all(np.isnan(group_covars_stacked) | (group_covars_stacked == np.floor(group_covars_stacked))):
+                self.covariates[group_name] = np.apply_along_axis(
+                    lambda x: x[~np.isnan(x)][0] if np.any(~np.isnan(x)) else np.nan, axis=0, arr=group_covars_stacked
+                )
```

Comment:
I think something like this would be faster:
```suggestion
                idx = np.isfinite(x)
                self.covariates[group_name] = np.where(np.any(idx, axis=1), x[:, np.argmax(idx, axis=1)], np.nan)
```